### PR TITLE
Add corner case to Degree Pearson Correlation Coefficient

### LINKED ISSUE #6913: ValueError crash in "nx.degree_pearson_correlation_coefficient" when input graph with no edges
Hello! I am so sorry to bother you to investigate again, but I have no idea about this probable issue. 
When testing NetworkX with a graph without edges, I found a strange crash in `nx.degree_pearson_correlation_coefficient`, with the following message:
```
Exception has occurred: ValueError
File "/home/qiuyang/.local/lib/python3.10/site-packages/networkx/algorithms/assortativity/correlation.py", line 156, in degree_pearson_correlation_coefficient
    x, y = zip(*xy)
ValueError: not enough values to unpack (expected 2, got 0)
```
The statement that triggers the crash is [here](https://github.com/networkx/networkx/blob/88097f7d7f798ec49eb868691dde77cf791a67ec/networkx/algorithms/assortativity/correlation.py#L158).
Since I am not sure whether `degree_pearson_correlation_coefficient` has a good definition in such graphs, it would be highly appreciated if you could help me further confirm and investigate this issue.
If there is no good definition, maybe we could throw a NetworkX Exception for users instead of the system exception.

Best regard,
Joye

### Step to Reproduce
Running the following Python code:
```Python
G = nx.DiGraph()
G.add_node(0)
G.add_node(1)
print(nx.degree_pearson_correlation_coefficient(G))
```
#### Result:
```
Exception has occurred: ValueError
```
### Environment
NetworkX 3.1
Python 3.10

### PR DESCRIPTION
**Pull request by vanshika230**
Closes #6913 
1. Add a corner to test for empty graphs. 
2. Add raises section in docs for empty graph.
3. Added test for the code added. The coverage is at 100%.
Please do let me know any changes to be made. Thankyou!

### CODE REVIEW

**Review by Schefflera-Arboricola — COMMENTED** [review]
(related corner case) Right now for a `Graph with 1 nodes and 1 edges`, `nx.degree_pearson_correlation_coefficient(G)` gives `ValueError: x and y must have length at least 2.`
```.py
>>> G = nx.Graph()
>>> G.add_node(1)
>>> G.add_edge(1,1)
>>> nx.degree_pearson_correlation_coefficient(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<class 'networkx.utils.decorators.argmap'> compilation 4", line 3, in argmap_degree_pearson_correlation_coefficient_1
  File "/workspaces/networkx/networkx/utils/backends.py", line 633, in __call__
    return self.orig_func(*args, **kwargs)
  File "/workspaces/networkx/networkx/algorithms/assortativity/correlation.py", line 168, in degree_pearson_correlation_coefficient
    return float(sp.stats.pearsonr(x, y)[0])
  File "/home/codespace/.local/lib/python3.10/site-packages/scipy/stats/_stats_py.py", line 4727, in pearsonr
    raise ValueError('x and y must have length at least 2.')
ValueError: x and y must have length at least 2.
```

File: networkx/algorithms/assortativity/tests/test_correlation.py [hunk 1]
```diff
@@ -62,6 +62,14 @@ def test_degree_assortativity_double_star(self):
         r = nx.degree_assortativity_coefficient(self.DS)
         np.testing.assert_almost_equal(r, -0.9339, decimal=4)
 
+    @pytest.mark.parametrize("G", (nx.Graph(), nx.DiGraph(), nx.MultiDiGraph()))
```
[the remark this answers is not in the document]
Comment (vanshika230) [reply networkx/algorithms/assortativity/tests/test_correlation.py:65]: I saw some other test files where all types were tested, if you feel this is unnecessary I can remove it do let me know!

**Comment by vanshika230** [comment]
> (related corner case) Right now for a `Graph with 1 nodes and 1 edges`, `nx.degree_pearson_correlation_coefficient(G)` gives `ValueError: x and y must have length at least 2.`
> 
> ```python
> >>> G = nx.Graph()
> >>> G.add_node(1)
> >>> G.add_edge(1,1)
> >>> nx.degree_pearson_correlation_coefficient(G)
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "<class 'networkx.utils.decorators.argmap'> compilation 4", line 3, in argmap_degree_pearson_correlation_coefficient_1
>   File "/workspaces/networkx/networkx/utils/backends.py", line 633, in __call__
>     return self.orig_func(*args, **kwargs)
>   File "/workspaces/networkx/networkx/algorithms/assortativity/correlation.py", line 168, in degree_pearson_correlation_coefficient
>     return float(sp.stats.pearsonr(x, y)[0])
>   File "/home/codespace/.local/lib/python3.10/site-packages/scipy/stats/_stats_py.py", line 4727, in pearsonr
>     raise ValueError('x and y must have length at least 2.')
> ValueError: x and y must have length at least 2.
> ```

I'll add for this, thankyou so much!

File: networkx/algorithms/assortativity/tests/test_correlation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply networkx/algorithms/assortativity/tests/test_correlation.py:65]: we do need parametrized tests sometimes, but here one type of a null graph would be enough to fail the test. Let me know what you think!

[the remark this answers is not in the document]
Comment (vanshika230) [reply networkx/algorithms/assortativity/tests/test_correlation.py:65]: Understood, thankyou. I'll change that.

**Review by rossbar — COMMENTED** [review]
Thanks @vanshika230 and @Schefflera-Arboricola for the review! I agree with @Schefflera-Arboricola that the additional corner cases should also be added.

Also, as a general comment, let's go ahead and switch all the new exceptions to `NetworkXError` - the other, more specific, exceptions (`NetworkXPointlessConcept` and `NetworkXUnfeasible`) have specialized meanings that don't necessarily map cleanly to these cases IMO!

**Comment by vanshika230** [comment]
> Thanks @vanshika230 and @Schefflera-Arboricola for the review! I agree with @Schefflera-Arboricola that the additional corner cases should also be added.
> 
> Also, as a general comment, let's go ahead and switch all the new exceptions to `NetworkXError` - the other, more specific, exceptions (`NetworkXPointlessConcept` and `NetworkXUnfeasible`) have specialized meanings that don't necessarily map cleanly to these cases IMO!

Noted, I have added the corner case told by @Schefflera-Arboricola , thankyou for that! Please let me know any further changes.

**Review by Schefflera-Arboricola — CHANGES_REQUESTED** [review]
Thanks @vanshika230 ! I've left a few comments below. LMK what u think.

File: networkx/algorithms/assortativity/correlation.py [hunk 2]
```diff
@@ -134,6 +134,13 @@ def degree_pearson_correlation_coefficient(G, x="out", y="in", weight=None, node
     r : float
        Assortativity of graph by degree.
 
+    Raises
+    ------
+    NetworkXError
+        If the graph 'G' is empty.
+    ValueError
```
[the remark this answers is not in the document]
Comment (vanshika230) [reply networkx/algorithms/assortativity/correlation.py:141]: Yes, I agree with you as it will appropriately point that error is related to graph properties and not the input values. Thankyou for the review! I'll make this change.

File: networkx/algorithms/assortativity/correlation.py [hunk 3]
```diff
@@ -134,6 +134,13 @@ def degree_pearson_correlation_coefficient(G, x="out", y="in", weight=None, node
...
+        If the graph has insufficient data to compute the correlation coefficient.
```
[the remark this answers is not in the document]
Comment (vanshika230) [reply networkx/algorithms/assortativity/correlation.py:142]: Yes, it does need to be more clear. I'll keep that in mind from now on. Thanks for reviewing!

**Comment by vanshika230** [comment]
> Thanks @vanshika230 ! I've left a few comments below. LMK what u think.

I have made the changes suggested in the review, Thankyou!

File: networkx/algorithms/assortativity/correlation.py [hunk 4]
```diff
@@ -154,8 +160,16 @@ def degree_pearson_correlation_coefficient(G, x="out", y="in", weight=None, node
     """
     import scipy as sp
 
+    if nx.is_empty(G):
+        raise nx.NetworkXError(
+            "Cannot compute correlation coefficient on an empty graph."
+        )
```
Comment (rossbar) [inline networkx/algorithms/assortativity/correlation.py:166]: Assuming my reasoning in the above comment is correct (please double-check!), I think the real condition that we want to catch here is that the graph *with self-loop edges removed* is not empty. One way to check that is by comparing the edge set to the set of self-loop edges, e.g. `set(G.edges) - set(nx.selfloop_edges(G))`.

However, this will also lead to subtle changes in the algorithm for the case where the graph with multiple nodes that *only* contain self-loop edges. Right now, this case gives:

```python
>>> G = nx.Graph([(1, 1), (2, 2), (3, 3)])
>>> nx.degree_pearson_correlation_coefficient(G)
ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
nan
```

whereas after the change this would raise an exception instead of producing the `nan`. This seems reasonable to me, but I'm no expert!

File: networkx/algorithms/assortativity/correlation.py [hunk 5]
```diff
@@ -134,6 +134,12 @@ def degree_pearson_correlation_coefficient(G, x="out", y="in", weight=None, node
     r : float
        Assortativity of graph by degree.
 
+    Raises
+    ------
+    NetworkXError
+        If the graph 'G' is empty.
+        If the graph 'G' has nodes with degrees less than two.
```
Comment (rossbar) [inline networkx/algorithms/assortativity/correlation.py:141]: This is not the condition that triggers the exception - nodes with degree less than 2 are fine (otherwise the `path_graph(4)` example in the docstring would fail, since the end nodes have degree 1!). IIUC, the corner case is related to graphs that have *only* a single self-loop edge.

**Review by rossbar — COMMENTED** [review]
As often happens, I think this one is a bit trickier than it appears at first glance!

File: networkx/algorithms/assortativity/correlation.py (hunk 5 above)
Comment (Schefflera-Arboricola) [reply networkx/algorithms/assortativity/correlation.py:141]: > IIUC, the corner case is related to graphs that have only a single self-loop edge.

we also get `ValueError: x and y must have length at least 2.` for 1 edge `DiGraph`s. And there probably might be some other cases too where the `len(x) < 2 or len(y) < 2` condition might be false.

```.py
import networkx as nx
G = nx.DiGraph([(1, 2)])
G1 = nx.Graph(G) 
print(nx.degree_pearson_correlation_coefficient(G1))
print(nx.degree_pearson_correlation_coefficient(G))

# output
"""
nan
/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:4781: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(stats.ConstantInputWarning(msg))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-14-73f922e088fa>](https://localhost:8080/#) in <cell line: 5>()
      3 G1 = nx.Graph(G)
      4 print(nx.degree_pearson_correlation_coefficient(G1))
----> 5 print(nx.degree_pearson_correlation_coefficient(G))
      6 xy = nx.node_degree_xy(G, x="out", y="in")
      7 print(xy)

3 frames
[/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py](https://localhost:8080/#) in pearsonr(x, y, alternative, method)
   4766 
   4767     if n < 2:
-> 4768         raise ValueError('x and y must have length at least 2.')
   4769 
   4770     x = np.asarray(x)

ValueError: x and y must have length at least 2.
"""
```

**Review by rossbar — COMMENTED** [review]
Gentle ping @vanshika230 This one's been stale for a while and I think there's still more work to do to nail down exactly the full set of corner cases.

**Comment by MridulS** [comment]
I'll go ahead and close this as this hasn't been touched in a while, thanks again for the PR!
